File: app/schemas/driver_shift.py
import json
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class StrictModel(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class ScenarioLevel(StrictModel):
    name: str = Field(min_length=1, max_length=50)
    threshold: int = Field(ge=0, le=1000000)
    benefits: str = Field(min_length=1, max_length=400)


class ScenarioTariff(StrictModel):
    id: str = Field(pattern=r"^[a-z0-9_-]{1,40}$")
    name: str = Field(min_length=1, max_length=60)
    available: bool = True
    reason: str = Field(default="", max_length=200)


class SupportStep(StrictModel):
    question: str = Field(min_length=5, max_length=600)
    options: list[str] = Field(min_length=2, max_length=4)
    correct: int = Field(ge=0, le=3)
    explanation: str = Field(min_length=5, max_length=600)

    @model_validator(mode="after")
    def valid_options(self):
        if self.correct >= len(self.options) or any(
            not 1 <= len(x.strip()) <= 160 for x in self.options
        ):
            raise ValueError("Проверьте варианты ответа и правильный ответ")
        return self


class DriverScenario(StrictModel):
    title: str = Field(default="Смена в Алматы", min_length=1, max_length=100)
    required_park: str = Field(default="itaxi", max_length=64)
    target_orders: int = Field(default=3, ge=1, le=10)
    fare: int = Field(default=1960, ge=100, le=100000)
    service_percent: float = Field(default=13.76, ge=0, le=40, allow_inf_nan=False)
    service_tax_percent: float = Field(default=16, ge=0, le=30, allow_inf_nan=False)
    wait_per_minute: int = Field(default=30, ge=0, le=1000)
    initial_balance: int = Field(default=0, ge=0, le=1000000)
    initial_points: int = Field(default=1485, ge=0, le=1000000)
    priority_base: int = Field(default=27, ge=0, le=100)
    priority_complete: int = Field(default=1, ge=0, le=10)
    priority_missed: int = Field(default=4, ge=0, le=30)
    priority_cancelled: int = Field(default=3, ge=0, le=30)
    offer_seconds: int = Field(default=30, ge=10, le=120)
    require_photo: bool = True
    require_documents: bool = True
    require_support: bool = True
    route_event: bool = True
    ratings: list[int] = Field(default=[0, 0, 2, 0, 148], min_length=5, max_length=5)
    levels: list[ScenarioLevel] = Field(
        default=[
            ScenarioLevel(name="Базовый", threshold=0, benefits="Доступ к учебным заказам"),
            ScenarioLevel(
                name="Эксперт",
                threshold=1000,
                benefits="Дополнительный приоритет и учебные предложения",
            ),
            ScenarioLevel(name="Профи", threshold=10000, benefits="Повышенный учебный приоритет"),
            ScenarioLevel(name="Чемпион", threshold=20000, benefits="Учебная скидка на комиссию"),
            ScenarioLevel(
                name="Гранд", threshold=50000, benefits="Все преимущества учебной программы"
            ),
        ],
        min_length=1,
        max_length=10,
    )
    tariffs: list[ScenarioTariff] = Field(
        default=[
            ScenarioTariff(id="economy", name="Эконом"),
            ScenarioTariff(id="comfort", name="Комфорт"),
            ScenarioTariff(
                id="comfort-plus",
                name="Комфорт+",
                available=False,
                reason="Учебный автомобиль не соответствует этой категории",
            ),
            ScenarioTariff(id="intercity", name="Межгород"),
            ScenarioTariff(
                id="kids",
                name="Детский",
                available=False,
                reason="Добавьте сведения о детском кресле в учебный профиль",
            ),
        ],
        min_length=1,
        max_length=20,
    )
    support_steps: list[SupportStep] = Field(
        default=[
            SupportStep(
                question="Водитель сообщает, что выплата не поступила. Что проверить сначала?",
                options=[
                    "Историю денег и статус выплаты",
                    "Сразу повторить вывод",
                    "Поменять автомобиль",
                ],
                correct=0,
                explanation=(
                    "Откройте Деньги → История транзакций → Выплата. Повторный вывод "
                    "не объясняет состояние первого запроса."
                ),
            ),
            SupportStep(
                question="Какую информацию приложить к учебному обращению?",
                options=[
                    "Пароль и код входа",
                    "Сумму, время и статус операции",
                    "Номер карты целиком",
                ],
                correct=1,
                explanation=(
                    "Для этого учебного кейса нужны сумма, время и статус выплаты. "
                    "Пароль, код входа и реквизиты не запрашиваются."
                ),
            ),
            SupportStep(
                question=(
                    "Поддержка подтвердила завершение учебной выплаты. Что сделать "
                    "после возврата?"
                ),
                options=[
                    "Проверить обновлённый статус в Деньгах",
                    "Повторить тот же вывод",
                    "Удалить профиль",
                ],
                correct=0,
                explanation=(
                    "Вернитесь в симулятор и откройте операцию: её статус должен "
                    "обновиться. Это завершает учебное обращение."
                ),
            ),
        ],
        min_length=1,
        max_length=10,
    )

    @model_validator(mode="after")
    def consistent(self):
        if any(x < 0 or x > 10000 for x in self.ratings) or not sum(self.ratings):
            raise ValueError("Введите количество оценок от 1 до 5 звёзд")
        if self.levels[0].threshold != 0 or any(
            a.threshold >= b.threshold for a, b in zip(self.levels, self.levels[1:], strict=False)
        ):
            raise ValueError("Первый уровень начинается с 0, остальные пороги возрастают")
        if len({x.id for x in self.tariffs}) != len(self.tariffs) or not any(
            x.available for x in self.tariffs
        ):
            raise ValueError("Тарифы должны иметь разные идентификаторы и хотя бы один доступный")
        return self


class ShiftStart(StrictModel):
    id: UUID
    mode: Literal["free", "assessment"]


class ShiftAction(StrictModel):
    request_id: UUID
    action: Literal[
        "visit",
        "setting",
        "tariff",
        "payment",
        "car_add",
        "car_select",
        "photo_step",
        "photo_submit",
        "photo_restart",
        "doc_sign",
        "provider",
        "wallet",
        "promo",
        "intercity_create",
        "intercity_book",
        "rental",
        "refuel",
        "learning",
        "work_mode",
        "online",
        "offline",
        "hint",
        "finish",
        "passenger",
        "route_change",
        "level_restore",
        "intercity_cancel",
    ]
    values: dict = Field(default_factory=dict)

    @model_validator(mode="after")
    def bounded(self):
        if len(json.dumps(self.values, ensure_ascii=False)) > 5000:
            raise ValueError("Слишком большой запрос")
        return self


class SupportCreate(StrictModel):
    id: UUID
    topic: Literal["payment", "passenger", "account"]

File: frontend/src/api/driverShift.ts
import { request } from "./client";
import { driverDeviceHeaders, type DriverState } from "./driver";
import type { NavigationSpec } from "./driverNavigation";

export interface SupportStep { question: string; options: string[]; correct: number; explanation: string }
export interface DriverScenario {
  real_location_required: boolean; arrival_radius: number; free_wait_seconds: number; boarding_seconds: number; fare_per_km: number; virtual_speed: number;
  title: string; required_park: string; target_orders: number; fare: number; service_percent: number; service_tax_percent: number;
  wait_per_minute: number; initial_balance: number; initial_points: number; priority_base: number; priority_complete: number;
  priority_missed: number; priority_cancelled: number; offer_seconds: number; require_photo: boolean; require_documents: boolean;
  require_support: boolean; route_event: boolean; ratings: number[];
  levels: { name: string; threshold: number; benefits: string }[];
  tariffs: { id: string; name: string; available: boolean; reason: string }[];
  support_steps?: SupportStep[];
}
export interface DriverPreferences {
  theme: "dark" | "light" | "system"; volume: number; vibration: boolean; hide_income: boolean; widgets: boolean;
  auto_arrive: boolean; auto_start: boolean; destination_marker: boolean; navigation: "internal" | "overview";
  location: "virtual" | "gps"; demand: boolean; traffic: boolean; bonus_zones: boolean; special_zones: boolean;
}
export interface LedgerEntry { id: string; title: string; amount: number; kind: string; order_id: string | null; status: "complete" | "pending" | "failed"; at: string; note: string }
export interface DriverMessage { id: string; channel: string; title: string; text: string; at: string; amount?: number; transaction_id?: string }
export interface DriverCar { id: string; brand: string; model: string; year: number; plate: string; status: string }
export interface IntercityOffer { id: string; origin: string; destination: string; date?: string; from_time?: string; to_time?: string; seats?: number; type?: string; price: number; status: string }
export interface ShiftResult { trips?: { id: string; origin: string; destination: string; fare: number; payment: string; navigation: NavigationSpec }[]; score: number | null; penalties: number; checks: { key: string; title: string; path: string; done: boolean; weight: number }[]; orders: number; target: number; seconds: number; errors: number; hints: number }
export interface DriverShift {
  is_preview?: boolean; content_id?: number | null;
  id: string; mode: "free" | "assessment"; config: DriverScenario; version: number; created_at: string; finished_at: string | null; result: ShiftResult | null;
  events: { id: string; action: string; at: string; details: Record<string, unknown> }[];
  data: {
    online: boolean; completed: number; missed: number; cancelled: number; hints: number; priority: number; points: number; seen: string[];
    photo_steps: number[]; photo_status: string; settings: DriverPreferences; tariffs: string[]; payment: "any" | "cash" | "card";
    cars: DriverCar[]; car_id: string; provider: string | null; documents: { id: string; title: string; period: string; status: string; signed_at: string | null }[];
    ledger: LedgerEntry[]; messages: DriverMessage[]; bookings: IntercityOffer[]; intercity_alerts: { text: string; at: string }[];
    rentals: { car: string; contact: string; status: string }[]; fuel: { liters: number; amount: number; at: string }[];
    promos: string[]; lessons: string[]; passenger_messages: { order_id: string; text: string; reply: string; at: string }[];
    work_mode: string; mode_address: string; support_completed: number; support_errors: number; support_checked_wallet: boolean; payout_problem: boolean;
    level_restored: boolean; points_history: { order_id: string; points: number; at: string }[]; rating_votes: number[];
    balance: number; available: number; reserved: number;
  };
}
export interface SupportCase { id: string; topic: string; status: string; step: number; total: number; created_at: string; finished_at: string | null; order_id: string | null }
export type ShiftActionName = "visit" | "setting" | "tariff" | "payment" | "car_add" | "car_select" | "photo_step" | "photo_submit" | "photo_restart" | "doc_sign" | "provider" | "wallet" | "promo" | "intercity_create" | "intercity_book" | "rental" | "refuel" | "learning" | "work_mode" | "online" | "offline" | "hint" | "finish" | "passenger" | "route_change" | "level_restore" | "intercity_cancel";
export type ShiftAct = (action: ShiftActionName, values?: Record<string, unknown>) => void;
export const driverShift = {
  results: (page: number, userId?: number) => request<{ total: number; items: { id: string; full_name: string; title: string; mode: string; completed: number; created_at: string; finished_at: string | null; result: ShiftResult | null }[] }>(`/api/v1/admin/learning/driver-results?page=${page}&size=20${userId ? `&user_id=${userId}` : ""}`),
  start: (json: { id: string; mode: "free" | "assessment"; content_id?: number }) => request<DriverState>("/api/v1/learning/driver/shifts", { method: "POST", json, headers: driverDeviceHeaders() }),
  act: ({ id, ...json }: { id: string; request_id: string; action: ShiftActionName; values: Record<string, unknown> }) => request<DriverState>(`/api/v1/learning/driver/shifts/${id}/action`, { method: "PUT", json, headers: driverDeviceHeaders() }),
  support: ({ shift_id, ...json }: { shift_id: string; id: string; topic: "payment" | "passenger" | "account" }) => request<{ case: SupportCase; url: string | null }>(`/api/v1/learning/driver/shifts/${shift_id}/support`, { method: "POST", json, headers: driverDeviceHeaders() }),
  config: () => request<DriverScenario>("/api/v1/admin/learning/driver-scenario"),
  saveConfig: (json: DriverScenario) => request<DriverScenario>("/api/v1/admin/learning/driver-scenario", { method: "PUT", json }),
};
